pkg/host-disk/host-disk.go

```diff
@@ -0,0 +1,80 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2018 Red Hat, Inc.
+ *
+ */
+
+package hostdisk
+
+import (
+	"os"
+	"os/exec"
+	"path"
+	"strconv"
+
+	k8sv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"kubevirt.io/kubevirt/pkg/api/v1"
+	"kubevirt.io/kubevirt/pkg/kubecli"
+	"kubevirt.io/kubevirt/pkg/precond"
+)
+
+const pvcBaseDir = "/var/run/kubevirt-private/vmi-disks"
+
+func getDiskImgPath(volumeName string) string {
+	return path.Join(pvcBaseDir, volumeName, "disk.img")
+}
+
+func calculateRawImgSize(quantity resource.Quantity) int64 {
+	// TODO: take fs overhead into account
```

Comment:
If you have the path of the mounted FS, you can do something like:

```
func dirMegabytesAvailable(path string) (uint64, error) {
       var stat syscall.Statfs_t
       err := syscall.Statfs(path, &stat)
       if err != nil {
               return 0, err
       }

       return (stat.Bavail * uint64(stat.Bsize) >> 20), nil
}
```

This will tell you what space is actually there on the FS, minus things like admin reserve (5% on ext, 0% of xfs).
I would not rely on the PVC size from the API, or at least sanity check it against the actual system FS available.

Reply:
> This will tell you what space is actually there on the FS, minus things like admin reserve (5% on ext, 0% of xfs).
> I would not rely on the PVC size from the API, or at least sanity check it against the actual system FS available.

@mlsorensen For me personally it is not really clear what to do in such a situation. I guess the correct thing would be to fail starting the vm. I thought about the root reservation which you also mentioned in the issue itself. It looks to me more like a bug in the storage provisioner than something the application should think of. What do you think?

Reply:
I think maybe this implementation has gone beyond what I had originally envisioned.  I had figured that the PVC would simply have a size, and under certain circumstances a disk,img would be created that matched the PVC's full size. In this case there would be no need to have HostDisk with its own capacity.

In this case it seems perhaps that we have a PVC of a predetermined size(say 100G), and we may be asking for a host disk of a different capacity, for example a 10G HostDisk to be created on it.  If this is the design, then yes, you could sanity check that the PVC's available space (via statfs) can hold it to avoid errors as the disk image fills up. I can think of a few issues that could arise if the intention is to share the PV for multiple disk images or data and only the sparse sizes are accounted for.

If the intention is not to share, then I wonder why HostDisk has its own size. Simply use the PVC size or detect it's usable space as above.

At any rate, please take or leave my comment as you wish. I was  just giving an idea of how one may implement a  "TODO: take fs overhead into account". I think I need to go back and understand the implementation a little better.

Reply:
> If the intention is not to share, then I wonder why HostDisk has its own size. Simply use the PVC size or detect it's usable space as above.

@mlsorensen I understand the confusion. Also lies within my responsibility since I asked @mfranczy to work into this direction. The `HostDisk` is in general an independent feature. So if you specify a `HostDisk` on the VM, then we would mount a HostPath and create the disk if requested with the specified size.

For the PVC initialization, the host-disk also works as a vehicle between virt-handler and virt-launcher to pass in the size which from the PVC. The rational behind this is that for virt-launcher it does not make any difference if the mounted path is coming from a PVC or from a hostDisk.

I find the `HostDisk` a useful missing feature and just found that it is a nice base for the PVC initialization work.

Reply:
@rmohr  @mlsorensen I have added new host disk volume source as independent feature.
If there is a ```PersistentVolumeClaim``` it's gonna be translated to ```HostDisk```,
and you are able to use ```HostDisk``` by itself definition.
Now i am writing detailed comment how it works exaclty...

@mlsorensen I have to update PR according to your comments about creating sparse file.

Reply:
@mlsorensen  regarding your comment:

>If you have the path of the mounted FS, you can do something like
...
>I would not rely on the PVC size from the API ...

I thought about that, but I have to rely on the PVC size from the API, if I wouldn't then I have the free space from the node. I used vmi-pvc.yaml from examples and checked how PVC is mounted into container. I guess we have to improve this?

Executed from the pvc container with the path of the mounted FS:
```
$ df -Ph /run/kubevirt-private/vmi-disks/pvcvolume
Filesystem                       Size  Used Avail Use% Mounted on
/dev/mapper/VolGroup00-LogVol00   38G  4.2G   34G  12% /run/kubevirt-private/vmi-disks/pvcvolume
```

Executed from the node
```
$ df -Ph /
Filesystem                       Size  Used Avail Use% Mounted on
/dev/mapper/VolGroup00-LogVol00   38G  4.2G   34G  12% /
```

Reply:
I see, so this is all related to using the node's filesystem for disk images.  In my world we only have PVCs that are independent block devices, and will eventually consume those directly, but in the mean time the option was to use it entirely. The above doesn't work if your mounted FS is not dedicated to the PVC.